### PR TITLE
include macros in Docs.undocumented_names

### PR DESCRIPTION
**Pull request by stevengj**
Fixes https://github.com/JuliaLang/julia/pull/52413#discussion_r1441068405

cc @jariji

### CODE REVIEW

**Comment by jariji** [comment]
LGTM

**Comment by stevengj** [comment]
CI failures look unrelated, should be good to merge @LilithHafner.

**Comment by LilithHafner** [comment]
This still has some strange behaviors/bugs
- undocumented_names() does not include deprecated names while undocumented_names(all=true) does
- it is documented to work based on export status but actually works based on publicity
- it doesn't return the names of infix operators

See https://github.com/JuliaLang/julia/pull/52413#issuecomment-1876266678 for an alternative semantic.

File: base/docs/Docs.jl [hunk 1]
```diff
@@ -681,13 +681,16 @@ end
...
+and `@macro` symbols are included. Names are returned in sorted order.
 
 See also: [`names`](@ref), [`Docs.hasdoc`](@ref), [`Base.isidentifier`](@ref).
 """
 function undocumented_names(mod::Module; all::Bool=false)
     filter!(names(mod; all)) do sym
-        !hasdoc(mod, sym) && Base.isidentifier(sym)
+        !hasdoc(mod, sym) && (Base.isidentifier(sym) ||
+            let str = string(sym)
+                startswith(str, "@") && Base.isidentifier(str[2:end])
```
Comment (stevengj) [inline base/docs/Docs.jl:692]: Use the single-char method:
```suggestion
                startswith(str, '@') && Base.isidentifier(str[2:end])
```

**Comment by stevengj** [comment]
Closed in favor of #52743

**Pull request closed without merging**